framework/auth/views.py

```diff
@@ -1153,8 +1156,8 @@ def external_login_email_post():
                 email=user.username
             )
             kind = 'success'
-            # 5. remove session
-            remove_session(session)
+            # 5. Clear session data
+            clear_external_first_login_anonymous_session_data(session)
```

Comment:
Previously, we were removing the sessions after sending the confirmation email. I understand that here we don't want to invoke `remove_session` because  `push_status_message` requires a valid session, so we'd only want to call the `clear_external_first_login_anonymous_session_data` here to pop the fields. But that will still leave in db/redis an orphaned session.

It probably doesn't matter because this session will automatically expire after 30 days. But does it make sense to `remove_session(session)` at the very end of this function before we return? I reckon that should be fine since we have already finished

Reply:
As discussed and locally tested, `session` is still used when the server tries to render the page. `pop_status_message` is called and passed to the `OsfWebRenderer`  as one of the globals.

See [`kwargs['data'] = get_globals`](https://github.com/CenterForOpenScience/osf.io/blob/fe96531ac9b3d48ced6025f23d03951b03ee5270/website/routes.py#L188) and [`'status': status.pop_status_messages(),`](https://github.com/CenterForOpenScience/osf.io/blob/develop/website/routes.py#L125).

Thus we have to keep this session. Here is the error log if we remove the session after the `push_status_message()`:

```
osf-web-1  | [werkzeug]  INFO: 172.18.0.1 - - [28/Mar/2023 15:48:46] "POST /external-login/email HTTP/1.1" 500 -                                                                      │t:0.028s]
osf-web-1  | Traceback (most recent call last):                                                                                                                                       │osf-api-1  | [28/Mar/2023 15:47:27] "POST /_/metrics/events/counted_usage/ HTTP/1.1" 201 0
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 2301, in __call__                                                                                           │osf-api-1  | /code/framework/auth/views.py changed, reloading.
osf-web-1  |     return self.wsgi_app(environ, start_response)                                                                                                                        │osf-api-1  | [django.utils.autoreload]  INFO: /code/framework/auth/views.py changed, reloading.
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 2287, in wsgi_app                                                                                           │osf-api-1  | [addons.osfstorage.settings.defaults]  WARNING: No migration settings loaded for OSFStorage, falling back to local dev. module 'website.settings' has no attribute 'MIGR
osf-web-1  |     response = self.handle_exception(e)                                                                                                                                  │ATION_ENV'
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1733, in handle_exception                                                                                   │osf-api-1  | [addons.osfstorage.settings]  WARNING: No local.py settings file found
osf-web-1  |     reraise(exc_type, exc_value, tb)                                                                                                                                     │osf-api-1  | [website.util.paths]  WARNING: Skipping load of "webpack-assets.json" in DEBUG_MODE.
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise                                                                                          │osf-api-1  | [addons.bitbucket.settings]  WARNING: No local.py settings file found
osf-web-1  |     raise value                                                                                                                                                          │osf-api-1  | [addons.dataverse.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 2284, in wsgi_app                                                                                           │osf-api-1  | [addons.dropbox.settings]  WARNING: No local.py settings file found
osf-web-1  |     response = self.full_dispatch_request()                                                                                                                              │osf-api-1  | [addons.figshare.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1807, in full_dispatch_request                                                                              │osf-api-1  | [addons.github.settings]  WARNING: No local.py settings file found
osf-web-1  |     rv = self.handle_user_exception(e)                                                                                                                                   │osf-api-1  | [addons.gitlab.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1710, in handle_user_exception                                                                              │osf-api-1  | [addons.googledrive.settings]  WARNING: No local.py settings file found
osf-web-1  |     reraise(exc_type, exc_value, tb)                                                                                                                                     │osf-api-1  | [addons.owncloud.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise                                                                                          │osf-api-1  | [addons.s3.settings]  WARNING: No local.py settings file found
osf-web-1  |     raise value                                                                                                                                                          │osf-api-1  | [addons.mendeley.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1805, in full_dispatch_request                                                                              │osf-api-1  | [addons.wiki.settings]  WARNING: No local.py settings file found
osf-web-1  |     rv = self.dispatch_request()                                                                                                                                         │osf-api-1  | [addons.zotero.settings]  WARNING: No local.py settings file found
osf-web-1  |   File "/usr/lib/python3.6/site-packages/flask/app.py", line 1791, in dispatch_request                                                                                   │osf-api-1  | [raven.contrib.django.client.DjangoClient]  INFO: Raven is not configured (logging is disabled). Please see the documentation for more information.
osf-web-1  |     return self.view_functions[rule.endpoint](**req.view_args)                                                                                                           │osf-api-1  | [website.app]  INFO: Sentry disabled; Flask's debug mode enabled
osf-web-1  |   File "/code/framework/routing/__init__.py", line 128, in wrapped                                                                                                       │osf-api-1  | Watching for file changes with StatReloader
osf-web-1  |     return renderer(data, **renderer_kwargs or {})                                                                                                                       │osf-api-1  | [django.utils.autoreload]  INFO: Watching for file changes with StatReloader
osf-web-1  |   File "/code/website/routes.py", line 197, in __call__                                                                                                                  │osf-api-1  | Performing system checks...
osf-web-1  |     resp = super(OsfWebRenderer, self).__call__(data, *args, **kwargs)                                                                                                   │osf-api-1  |
osf-web-1  |   File "/code/framework/routing/__init__.py", line 336, in __call__                                                                                                      │osf-api-1  | System check identified no issues (0 silenced).
osf-web-1  |     rendered = self.render(data, redirect_url, *args, **kwargs)                                                                                                          │osf-api-1  | March 28, 2023 - 15:48:26
osf-web-1  |   File "/code/framework/routing/__init__.py", line 605, in render                                                                                                        │osf-api-1  | Django version 3.2.17, using settings 'api.base.settings'
osf-web-1  |     extra_data = self.data if isinstance(self.data, dict) else self.data()                                                                                               │osf-api-1  | Starting development server at http://0.0.0.0:8000/
osf-web-1  |   File "/code/website/routes.py", line 125, in get_globals                                                                                                               │osf-api-1  | Quit the server with CONTROL-C.
osf-web-1  |     'status': status.pop_status_messages(),                                                                                                                              │osf-api-1  | [28/Mar/2023 15:48:33] "GET /v2/ HTTP/1.1" 200 1063
osf-web-1  |   File "/code/framework/status/__init__.py", line 67, in pop_status_messages                                                                                             │osf-api-1  | [28/Mar/2023 15:48:36] "GET /v2/?version=latest HTTP/1.1" 200 1063
osf-web-1  |     messages = current_session.get('status', None)                                                                                                                       │osf-api-1  | [28/Mar/2023 15:48:38] "GET /v2/status/ HTTP/1.1" 200 46
osf-web-1  | AttributeError: 'NoneType' object has no attribute 'get'
```

Reply:
There is a more sophisticated solution but that requires we update the `UserSessionMap` to keep track of such sessions and do one of the followings
* Remove them during our daily cron job
* Remove them when user click the confirmation link
* Remove after rendering before sending back the response